File: musicNotification/src/main/java/com/xzh/musicnotification/notification/BaseMusicNotification.java
package com.xzh.musicnotification.notification;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Build;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.KeyEvent;

import com.alibaba.fastjson.JSONObject;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DecodeFormat;
import com.bumptech.glide.request.target.Target;
import com.xzh.musicnotification.Global;
import com.xzh.musicnotification.service.PlayServiceV2;
import com.xzh.musicnotification.utils.Utils;

import java.lang.ref.WeakReference;

import io.dcloud.PandoraEntryActivity;

public abstract class BaseMusicNotification {
    protected int NOTIFICATION_ID = 0x111;
    protected static final String CHANNEL_ID = "music_id_audio";
    protected static final String CHANNEL_NAME = "music_name_audio";

    protected static JSONObject mConfig;
    protected static boolean showFavour;
    protected boolean isPlay;
    protected JSONObject songInfo;
    protected WeakReference<Context> mContext;
    protected Notification mNotification;
    protected MediaSessionCompat mMediaSession;
    protected NotificationManager mNotificationManager;

    @SuppressLint("UnspecifiedImmutableFlag")
    protected PendingIntent getContentIntent(String path) {
        Intent intent = new Intent(mContext.get(), PandoraEntryActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        if (path != null) {
            intent.putExtra(Global.KEY_PATH, path);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return PendingIntent.getActivity(mContext.get(), 0, intent, PendingIntent.FLAG_IMMUTABLE);
        } else {
            return PendingIntent.getActivity(mContext.get(), 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        }
    }

    public static void init(JSONObject config) {
        mConfig = config;
    }

    /**
     * 创建Notification,
     *
     * @param service BaseMusicNotification.NotificationHelperListener
     */
    public void initNotification(Service service) {
        mContext = new WeakReference<>(service);
        NOTIFICATION_ID++;

        mMediaSession = new MediaSessionCompat(service, BaseMusicNotification.CHANNEL_ID);
        mMediaSession.setActive(true);
        mMediaSession.setMetadata(new MediaMetadataCompat.Builder().build());
        mMediaSession.setPlaybackState((new PlaybackStateCompat.Builder()).setActions(PlaybackStateCompat.ACTION_PLAY
                | PlaybackStateCompat.ACTION_PLAY_PAUSE
                | PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID
                | PlaybackStateCompat.ACTION_PAUSE
                | PlaybackStateCompat.ACTION_SKIP_TO_NEXT
                | PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS).build());
        mMediaSession.setCallback(new MediaSessionCompat.Callback() {
            @Override
            public boolean onMediaButtonEvent(Intent intent) {
                KeyEvent keyEvent = intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT);
                if (keyEvent.getAction() == 0) {
                    JSONObject data = new JSONObject();
                    data.put("type", Global.MEDIA_BUTTON);
                    data.put("keyCode", keyEvent.getKeyCode());
                    ((PlayServiceV2) mContext.get()).fireGlobalEventCallback(Global.EVENT_MUSIC_MEDIA_BUTTON, data);
                }
                return true;
            }
        });
        createNotification();
    }

    public abstract void createNotification();

    protected abstract void updateNotification();

    /**
     * 更新 Notification 信息
     *
     * @param options 歌曲信息
     */
    public void updateSong(JSONObject options) {
        songInfo = options;
        updateNotification();
    }

    /**
     * 切换播放状态
     *
     * @param isPlay 播放状态
     */
    public void playOrPause(boolean isPlay) {
        this.isPlay = isPlay;
        updateNotification();
    }

    /**
     * 切换搜藏状态
     *
     * @param favourite 搜藏状态
     */
    public void favour(boolean favourite) {
        if(songInfo != null) songInfo.put(Global.KEY_FAVOUR, favourite);
        updateNotification();
    }

    public void cancel() {
        if (mNotificationManager != null) mNotificationManager.cancel(NOTIFICATION_ID);
    }

    public static void setShowFavour(boolean show) {
        showFavour = show;
    }

    protected void generateGlide(Context context, Target<Bitmap> target, String picUrl, float width, float height) {
        Glide.with(context)
                .asBitmap()
                .load(picUrl)
                .sizeMultiplier(0.8f)
                .format(DecodeFormat.PREFER_RGB_565)
                .override(Utils.dip2px(width), Utils.dip2px(height))
                .into(target);
    }
}

File: musicNotification/src/main/java/com/xzh/musicnotification/notification/BaseMusicNotification.kt
package com.xzh.musicnotification.notification

import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.os.Build
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.KeyEvent
import com.alibaba.fastjson.JSONObject
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DecodeFormat
import com.bumptech.glide.request.target.Target
import com.xzh.musicnotification.Global
import com.xzh.musicnotification.utils.Utils
import io.dcloud.PandoraEntryActivity
import java.lang.ref.WeakReference


abstract class BaseMusicNotification {

    companion object {
        internal const val CHANNEL_ID = "music_id_audio"

        internal const val CHANNEL_NAME = "music_name_audio"
        internal var mConfig: JSONObject? = null

        internal var showFavour = false
    }

    var iD = 0x111
        protected set

    @JvmField
    protected var position = 0L

    @JvmField
    protected var isPlay = false

    @JvmField
    protected var songInfo: MutableMap<String, Any?>? = null

    @JvmField
    protected var mContext: WeakReference<Context>? = null

    @JvmField
    protected var mNotification: Notification? = null

    @JvmField
    protected var mMediaSession: MediaSessionCompat? = null

    @JvmField
    protected var mNotificationManager: NotificationManager? = null

    @JvmField
    protected var mMusicEventListener: OnMusicEventListener? = null

    @JvmField
    protected var mPlaybackStateBuilder: PlaybackStateCompat.Builder? = null

    @SuppressLint("UnspecifiedImmutableFlag")
    protected fun getContentIntent(path: String?): PendingIntent {
        val intent = Intent(mContext!!.get(), PandoraEntryActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        if (path != null) {
            intent.putExtra(Global.KEY_PATH, path)
        }
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            PendingIntent.getActivity(mContext!!.get(), 0, intent, PendingIntent.FLAG_IMMUTABLE)
        } else {
            PendingIntent.getActivity(
                mContext!!.get(),
                0,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT
            )
        }
    }

    /**
     * 创建Notification,
     *
     * @param service BaseMusicNotification.NotificationHelperListener
     */
    fun initNotification(service: Service, listener: OnMusicEventListener) {
        mContext = WeakReference(service)
        mMusicEventListener = listener;
        iD += 1
        mMediaSession = MediaSessionCompat(service, CHANNEL_ID)
        mMediaSession!!.isActive = true
//        mMediaSession!!.setMetadata(MediaMetadataCompat.Builder().build())

        mPlaybackStateBuilder = PlaybackStateCompat.Builder()
            .setActions(
//            PlaybackStateCompat.ACTION_STOP or
                PlaybackStateCompat.ACTION_PAUSE or
                        PlaybackStateCompat.ACTION_PLAY or
                        PlaybackStateCompat.ACTION_REWIND or
                        PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or
                        PlaybackStateCompat.ACTION_SKIP_TO_NEXT or
                        PlaybackStateCompat.ACTION_FAST_FORWARD or
//                        PlaybackStateCompat.ACTION_SET_RATING or
                        PlaybackStateCompat.ACTION_SEEK_TO or
                        PlaybackStateCompat.ACTION_PLAY_PAUSE
//                        PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID or
//                        PlaybackStateCompat.ACTION_PLAY_FROM_SEARCH or
//                        PlaybackStateCompat.ACTION_SKIP_TO_QUEUE_ITEM or
//                        PlaybackStateCompat.ACTION_PLAY_FROM_URI or
//                        PlaybackStateCompat.ACTION_PREPARE or
//                        PlaybackStateCompat.ACTION_PREPARE_FROM_MEDIA_ID or
//                        PlaybackStateCompat.ACTION_PREPARE_FROM_SEARCH or
//                        PlaybackStateCompat.ACTION_PREPARE_FROM_URI or
//                        PlaybackStateCompat.ACTION_SET_REPEAT_MODE or
//                    PlaybackStateCompat.ACTION_SET_SHUFFLE_MODE
//                        PlaybackStateCompat.ACTION_SET_CAPTIONING_ENABLED or
//                        PlaybackStateCompat.ACTION_SET_PLAYBACK_SPEED
            )

        // 使用新的播放状态更新MediaSessionCompat实例
        mMediaSession?.setPlaybackState(mPlaybackStateBuilder?.build())

        mMediaSession?.setCallback(object : MediaSessionCompat.Callback() {
            override fun onMediaButtonEvent(intent: Intent): Boolean {
                val keyEvent: KeyEvent? =
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                        intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT, KeyEvent::class.java)
                    } else {
                        intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT)
                    }
                if (keyEvent?.action == 0) {
                    mMusicEventListener?.onMediaButtonEvent(keyEvent.keyCode)
                }
                return true
            }

            override fun onPlay() {
                mMusicEventListener?.onPlay()
            }

            override fun onPause() {
                mMusicEventListener?.onPause()
            }

            override fun onSkipToNext() {
                mMusicEventListener?.onSkipToNext()
            }

            override fun onSkipToPrevious() {
                mMusicEventListener?.onSkipToPrevious()
            }

            override fun onSeekTo(pos: Long) {
                this@BaseMusicNotification.setPosition(pos)
                // 设置播放状态为正在播放，并设置媒体播放的当前位置
                mMediaSession!!.setPlaybackState(
                    mPlaybackStateBuilder
                        ?.setState(
                            if (this@BaseMusicNotification.isPlay) PlaybackStateCompat.STATE_PLAYING else PlaybackStateCompat.STATE_STOPPED,
                            pos, 1.0F
                        )
                        ?.build()
                )
                mMusicEventListener?.onSeekTo(pos)
            }
        })
    }

    abstract fun createNotification()
    protected abstract fun updateNotification()

    /**
     * 更新 Notification 信息
     *
     * @param options 歌曲信息
     */
    fun updateSong(options: MutableMap<String, Any?>) {
        songInfo = options
        updateNotification()
    }

    /**
     * 切换播放状态
     *
     * @param isPlay 播放状态
     */
    fun playOrPause(isPlay: Boolean) {
        this.isPlay = isPlay
        updateNotification()
    }

    @JvmName("setBasePosition")
    fun setPosition(position: Long) {
        this.position = position
    }

    /**
     * 切换搜藏状态
     *
     * @param favourite 搜藏状态
     */
    fun favour(favourite: Boolean) {
        if (songInfo != null) songInfo!![Global.KEY_FAVOUR] = favourite
        updateNotification()
    }

    fun cancel() {
        if (mNotificationManager != null) {
            mNotification = null
            mNotificationManager!!.cancel(iD)
        }
    }

    protected fun generateGlide(
        context: Context?,
        target: Target<Bitmap>,
        picUrl: String?,
        width: Float,
        height: Float
    ) {
        Glide.with(context!!)
            .asBitmap()
            .load(picUrl)
            .sizeMultiplier(0.5f)
            .format(DecodeFormat.PREFER_RGB_565)
            .override(Utils.dip2px(width), Utils.dip2px(height))
            .into(target)
    }

    interface OnMusicEventListener {
        fun onMediaButtonEvent(keyCode: Int)
        fun onPlay()
        fun onPause()
        fun onSkipToNext()
        fun onSkipToPrevious()
        fun onSeekTo(pos: Long)
    }
}